modules/tensor_mechanics/src/actions/ProjectedStatefulMaterialStorageAction.C

```diff
@@ -0,0 +1,235 @@
+              Utility::string_to_enum<FEFamily>(params.get<MooseEnum>("family"))}),
+    _var_type(AddVariableAction::variableType(_fe_type)),
+    _pomps_prefix("_pomps_")
+{
+}
+
+void
+ProjectedStatefulMaterialStorageAction::addRelationshipManagers(
+    Moose::RelationshipManagerType input_rm_type)
+{
+  auto params = _factory.getValidParams("ProjectedStatefulMaterialNodalPatchRecovery");
+  addRelationshipManagers(input_rm_type, params);
+}
+
+void
+ProjectedStatefulMaterialStorageAction::processComponent(const std::string & prop_name,
+                                                         std::vector<unsigned int> idx,
+                                                         std::vector<VariableName> & vars,
+                                                         bool is_ad)
+{
+  const std::string pomps_prefix = "_pomps_";
+  auto name = [&](const std::string & type)
+  {
+    auto name = _pomps_prefix + type + "_" + prop_name;
+    if (idx.size())
+      name += "_" + Moose::stringify(idx, "_");
+    return name;
+  };
+
+  const auto var_name = name("var");
+  vars.push_back(var_name);
+  const auto uo_name = name("uo");
+
+  if (_current_task == "setup_projected_properties")
+  {
+    // add the AuxVars for storage
+    auto params = _factory.getValidParams(_var_type);
+    params.applyParameters(parameters());
+    params.set<std::vector<OutputName>>("outputs") = {"none"};
+    _problem->addAuxVariable(_var_type, var_name, params);
+
+    // use nodal patch recovery for lagrange
+    if (_fe_type.family == LAGRANGE)
+    {
+      // nodal variables require patch recovery (add user object)
+      const auto & type_name = is_ad ? "ADProjectedStatefulMaterialNodalPatchRecovery"
+                                     : "ProjectedStatefulMaterialNodalPatchRecovery";
+      auto params = _factory.getValidParams(type_name);
+      params.applySpecificParameters(parameters(), {"block"});
+      params.set<std::vector<unsigned int>>("component") = idx;
+      params.set<MaterialPropertyName>("property") = prop_name;
+      params.set<MooseEnum>("patch_polynomial_order") = _order;
+      params.set<ExecFlagEnum>("execute_on") = {EXEC_INITIAL, EXEC_TIMESTEP_END};
+      params.set<bool>("force_preaux") = true;
+      _problem->addUserObject(type_name, uo_name, params);
+    }
+  }
+
+  if (_current_task == "add_aux_kernel")
+  {
+    // use nodal patch recovery for lagrange
+    if (_fe_type.family == LAGRANGE)
+    {
+      // nodal variables require patch recovery (add aux kernel)
+      auto params = _factory.getValidParams("NodalPatchRecoveryAux");
+      params.applySpecificParameters(parameters(), {"block"});
+      params.set<AuxVariableName>("variable") = var_name;
+      params.set<UserObjectName>("nodal_patch_recovery_uo") = uo_name;
+      params.set<ExecFlagEnum>("execute_on") = {EXEC_INITIAL, EXEC_TIMESTEP_END};
+      _problem->addAuxKernel("NodalPatchRecoveryAux", name("aux"), params);
+    }
+    else
```

Comment:
I think that this is ok for a start. Eventually I would prefer it if we don't have any special casing. The special casing would happen in the aux kernel ... which ideally would inherit from `ProjectionAux`. And if `ProjectionAux` doesn't supply all the capability we need (e.g. a nodal patch recovery for projecting into a Lagrange family), then we enhance it so it does

Reply:
I know that currently `ProjectionAux` takes a coupled variable as an input, but we could factor the important code into a base class. All that matters is registering the property we are projecting as being discontinuous

Reply:
I suppose we could do that with the proper level of ghosting (point neighbors). Right now we process data in an element user object (where we assemble the A matrix and b vector for the least square fits). That UO explicitly communicates element data from point neighbors on neighboring processors, but with algebraic ghosting we could evaluate it all locally. My concern would then only be performance. We only assemble each element matrix once and reuse it at every node on that element when looping over the nodes. To do everything in the AuxKernel we'd assemble those matrices repeatedly.

Reply:
You could have the aux kernel optionally take your user object that does it the more optimized way, else it does nodal patch recovery the more expensive way or something else that's not nodal patch recovery (something like it does now)